File: extensions/youtube/src/main/java/app/morphe/extension/youtube/patches/TranscriptPatch.java
package app.morphe.extension.youtube.patches;

import static app.morphe.extension.shared.spoof.SpoofAppVersionPatch.isSpoofingToLessThan;

import app.morphe.extension.youtube.settings.Settings;

@SuppressWarnings("unused")
public class TranscriptPatch {
    private static final boolean OVERRIDE_TRANSCRIPT_APP_VERSION = Settings.FIX_TRANSCRIPT.get()
            && !isSpoofingToLessThan("20.05.00");

    /**
     * Injection point.
     */
    public static String getTranscriptAppVersionOverride(String version) {
        return OVERRIDE_TRANSCRIPT_APP_VERSION
                ? "20.05.46"
                : version;
    }
}

File: patches/src/main/kotlin/app/morphe/patches/youtube/layout/captions/TranscriptPatch.kt
package app.morphe.patches.youtube.layout.captions

import app.morphe.patcher.patch.bytecodePatch
import app.morphe.patches.shared.misc.settings.preference.SwitchPreference
import app.morphe.patches.youtube.misc.contexthook.Endpoint
import app.morphe.patches.youtube.misc.contexthook.addClientVersionHook
import app.morphe.patches.youtube.misc.contexthook.clientContextHookPatch
import app.morphe.patches.youtube.misc.extension.sharedExtensionPatch
import app.morphe.patches.youtube.misc.playservice.versionCheckPatch
import app.morphe.patches.youtube.misc.settings.settingsPatch

private const val EXTENSION_CLASS =
    "Lapp/morphe/extension/youtube/patches/TranscriptPatch;"

internal val transcriptPatch = bytecodePatch(
    description = "Add an option to fix an issue where transcript is unavailable due to a precondition check failure.",
) {
    dependsOn(
        sharedExtensionPatch,
        settingsPatch,
        versionCheckPatch,
        clientContextHookPatch,
    )

    execute {
        settingsMenuCaptionGroup.add(
            SwitchPreference("morphe_fix_transcript")
        )

        addClientVersionHook(
            Endpoint.TRANSCRIPT,
            "$EXTENSION_CLASS->getTranscriptAppVersionOverride(Ljava/lang/String;)Ljava/lang/String;",
        )
    }
}
